README.md
[![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat-square)](http://bioconda.github.io/recipes/minvar/README.html)
[![codebeat badge](https://codebeat.co/badges/eff65e46-db50-4ad7-92ff-19a6a67bcaa2)](https://codebeat.co/projects/github-com-ozagordi-minvar-master)


# MinVar: automatic detection of drug-resistance mutations in HIV-1

MinVar is a command-line tool to discover mutations conferring drug resistance
in HIV-1 and HCV populations using deep sequencing data.

----

#### The simplest example

    [user@host ~]$ minvar -f sample_file.fastq
    ... a few minutes later ...
    [user@host ~]$ column -t -s ',' merged_muts_drm_annotated.csv
    gene      pos  mut  freq    category
    ...
    RT        238  T    1.0     NNRTI
    RT        250  N    0.9547  unannotated
    RT        272  P    1.0     unannotated
    RT        293  V    1.0     unannotated
    RT        297  A    1.0     unannotated
    RT        333  D    0.9384  unannotated
    RT        333  E    0.0354  unannotated
    RT        335  C    1.0     unannotated
    protease  10   P    0.0223  Other
    protease  10   Q    0.0185  Other
    protease  10   S    0.0741  Other
    protease  10   T    0.0468  Other
    protease  10   V    0.5948  PIMinor
    protease  11   L    1.0     PIMinor
    protease  13   V    1.0     unannotated
    protease  14   R    1.0     unannotated
    protease  15   V    0.7143  unannotated
    protease  20   T    1.0     PIMinor
    protease  32   I    1.0     PIMajor
    ...

## Important features

- MinVar is an _opinionated_ software: it just takes a fastq file as input and
  does not ask the standard user to set any parameter at run time. Nevertheless,
  the experienced user/developer can easily change some of its settings in the
  source code.
- It has been tested with HIV-1 on both Illumina MiSeq and Roche/454 sequencing
  reads. HCV has been tested on MiSeq only.
- It uses state-of-the-art third tools to filter, recalibrate, and align reads
  and to call variants.
- Finally, single nucleotide variants are phased at codon level and
  amino acid mutations are called and annotated.
- HIV-1 drug-resistance mutations are annotated according to
  [Stanford HIV Drug Resistance Database (HIVDB)](https://hivdb.stanford.edu).
- The annotated mutations are saved in a csv file (see example above) and also
  included in a report in markdown format that is finally converted to PDF.
- The PDF report can be customized by adding contact information specified in the
  file `~/.minvar/contact.ini` with the following syntax (only change what comes
  after the `=` sign)

```
[contact]
unit = name_of_your_unit_here
phone = phone_number
fax = fax_number
email = your_unit@your_company
logo = filename_without_extension
```

The logo file in pdf format must be present in the same directory. In other words,
if we want to use the file `~/.minvar/company_logo_bw.pdf`, then in the INI file we
will write `logo = company_logo_bw`.

## Documentation

See the official [documentation](http://minvar.readthedocs.io/en/latest/).

API documentation can be created by cloning this repo, `cd`-ing into
`apidoc` and running `make html`.

#### Citation

MinVar (version 1, HIV-1 support only) has been introduced and validated in  
Huber, Metzner _et al._, (2017) MinVar: A rapid and versatile tool for HIV-1
drug resistance genotyping by deep sequencing _Journal of virological methods_
240:7-13, [doi:10.1016/j.jviromet.2016.11.008](http://dx.doi.org/10.1016/j.jviromet.2016.11.008)


## Output files

### Created by `prepare.py`

- `subtype_evidence.csv` percent of reads best aligned to each subtype (or
  genotype),
- `subtype_ref.fasta` references of the subtype identified,
- `cns_final.fasta`: sample consensus created by iteratively aligning reads and
  writing variants into the sequence,

### Created by `callvar.py`

- `hq_2_cns_final_recal.bam` sorted bam alignment of reads to the consensus
  sequence, recalibrated with either GATK or lofreq (indels only),
- `hq_2_cns_final_recal.vcf` VCF file of mutations found on reads with respect
  to consensus in `cns_final.fasta`.

### Created by `annotate.py`

- `merged_mutations_nt.csv` a list of all variants observed at single positions,
- `max_freq_muts_aa.csv` the amminoacid found at maximum frequency at each codon,
- `final.csv` mutations at amminoacid level with indication of
  the gene, the position on the gene, wild type and frequency

### Created by `reportdrm.py`

- `merged_muts_drm_annotated.csv` is the join of `final.csv` with the annotation
  of DRM/RAS,
- `report.md` and `report.pdf` final report with subtye estimate based on
  alignment of reads to different references and tables with mutations. The
  pdf report is created from the template `minvar/db/template.tex`.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added
- Coverage information is added to the merged_mutations_nt.csv file.

### Fixed
- 

## [v2.2.1] - 2018-10-17

### Changed
- Ambiguity score is now computed on PRRT up to aa 335.

### Fixed
- Bugfix: unannotated mutations were not always reported in the tables (issue #25).
- Reference and consensus protein sequences having the same end would make `annotate.py` fail, fixed.

## [v2.2] - 2017-02-05

### Added
- Report now includes ambiguity score derived from ambiguous consensus sequence.
- Consensus sequence with ambiguous bases is saved into `cns_ambiguous.fasta`.
- Cells of drug resistance report are colour coded.
- The report includes levels of drug resistance inferred by [HIVdb](https://hivdb.stanford.edu/hivdb/by-mutations/).

### Changed
- Recombinant sequences are checked for also with HIV data.
- Threshold for sierra is read from the INI file `~/.minvar/runpars.ini`
- Minor changes in final report.
- `masterComments` files with DRMs lists are updated to version [8.4](https://hivdb.stanford.edu/page/algorithm-updates/)
- Both full drug name and abbreviation are written.
- Subtype inference only reports top hit, subtype names are now clearer.


## [v2.1.3] - 2017-01-18
- Date added on every page of the report, major mutations are in bold.
- Bugfix: minvar now works evene when neither HCV nor HIV reads are present.
- Bugfix: `find_subtype` failed when reads from both HIV and HCV were present.
- Bugfix: when no RAS mutation is found, `reportdrm` failed.
- Bugfix: when the frequency of a mutation is 0.5/0.5, `cns_max_freq` was longer than the reference.
- Problems with an assert when all reads have indels: removed the assert, `phase_mutations` has limitations.

## [v2.1.2] - 2017-12-19
- More information on the run included in the report.

## [v2.1.1] - 2017-12-19
- Fixed a bug showing up when sample consensus starts with a stop codon.

## [v2.1] - 2017-12-18

### Added
- Support for non-overlapping amplicons.

## [v2.0] - 2017-12-13

### Added
- Support for HCV.
- Report in pdf now styled as article with information on footer of each page.
- Contact information and a logo can be specified with a INI file in `~/.minvar`
- `minvar` can be called as a module or as command line program.

### Changed
- Improved generation of sample consensus.
- Widespread use of `shlex`, _i.e._ processes are not called via shell.
- Reorganization of the package structure:
  - code is now in `src/minvar`;
  - data are now in `src/minvar/db`;
  - data now include plenty of resources for HCV.
- Import system has been changed.
- Tests have been adapted (coverage is still very low).
- Code quality via flake8, pep8, isort and pep257.

## [v1.2b] - 2017-07-08
### Changed
- `sort` in `pandas` has been replaced by `sort_values`.
- Latex template needs to define `\highlight`.

## [v1.2a3] - 2017-06-13
### Added
- MinVar is now on bioconda.

## [v1.2alpha] - 2017-06-07
### Added
- `minvar -v` now returns version

## [v1.1] - 2017-03-08
### Added
- Switch `-r` added to turn on/off the calibration with GATK.
- Documentation.
- Ansible instructions.

## [v1.0.1] - 2016-09-20
Production version, tested on 454 and Miseq and used in the manuscript.


minvar/reportdrm.py
#!/usr/bin/env python3
'''Parse the different output files and write a report in markdown,
then convert it to pdf with markdown.
'''

import sys
import os
import csv
# from pkg_resources import resource_string
# foo_config = resource_string(__name__, 'foo.conf')
import pandas as pd

#from common import acc_numbers, hcv_map

# aminoacid one-letter code
aa_set = set('GPAVLIMCFYWHKRQNEDST')

# amminoacid sequences from files in db directory
dn_dir = os.path.dirname(__file__)
db_dir = os.path.abspath(os.path.join(dn_dir, 'db'))

# prot = \
#     list(SeqIO.parse(os.path.join(db_dir, 'protease.faa'), 'fasta'))[0]
# RT = list(SeqIO.parse(os.path.join(db_dir, 'RT.faa'), 'fasta'))[0]
# integrase = \
#     list(SeqIO.parse(os.path.join(db_dir, 'integrase.faa'), 'fasta'))[0]

def parse_drm():
    '''Parse drug resistance mutations listed in files db/*Variation.txt'''

    df_list = []
    for gene, drm_file_name in [('protease', 'masterComments_PI.txt'),
                                ('RT', 'masterComments_RTI.txt'),
                                ('integrase', 'masterComments_INI.txt')]:
        drm_file = os.path.join(db_dir, drm_file_name)
        try:
            d1 = pd.read_table(drm_file, header=0,
                               names=['pos', 'mut', 'category', 'commented',
                                      'comment'])
        except:  # integrase does not have commented column
            d1 = pd.read_table(drm_file, header=0,
                               names=['pos', 'mut', 'category', 'comment'])
        gs = pd.Series([gene] * len(d1))
        d1['gene'] = gs
        df_list.append(d1)
    df = pd.concat(df_list)
    return df

# def parse_region(hap1):
#     '''Whether the region comes from protease, RT or integrase'''
#
#     import Alignment
#     matched_res = []
#
#     for gs in [prot, RT, integrase]:
#         alout = '%s.tmp' % gs.id
#         Alignment.needle_align('asis:%s' % hap1, 'asis:%s' % str(gs.seq),
#                                alout)
#         ald = Alignment.alignfile2dict([alout])
#         os.remove(alout)
#         ali = ald['asis']['asis']
#         ali.summary()
#         ratio1 = float(ali.ident) / (ali.ident + ali.mismatches)
#         if ratio1 > 0.75:
#             matched_res.append(gs.id)
#
#     return matched_res


def write_header(handle, subtype_file=None, org_found=None):
    '''Write header to a file in markdown format'''
    from operator import itemgetter
    md_header = 'Drug resistance mutations detected by NGS sequencing'
    md_header += '\n' + '=' * len(md_header) + '\n\n'
    md_header += 'Subtype inference with blast\n'
    md_header += '----------------------------\n'
    md_header += '|     subtype     | support [%] |\n'
    md_header += '|:{:-^15}:|:{:-^11}:|\n'.format('', '')
    if subtype_file:
        save_freq = {}
        with open(subtype_file) as csvfile:
            spamreader = csv.reader(csvfile, delimiter=',')
            for mtype, freq in spamreader:
                int_freq = int(round(100 * float(freq), 0))
                if int_freq >= 1:
                    save_freq[mtype] = save_freq.get(mtype, 0) + int_freq
        for k, v in sorted(save_freq.items(), key=itemgetter(1),
                           reverse=True):
            md_header += '|{: ^17}|{: ^13}|\n'.format(k, v)
    if org_found == 'HIV':
        md_header += '''

Parsing mutations
-----------------

The list of mutations was downloaded from HIVdb and includes:

- xyz positions on protease
- zyx positions on RT
- abc positions on integrase.
'''
    print(md_header, file=handle)


def parse_com_line():
    '''argparse, since optparse deprecated starting from python 2.7'''
    import argparse

    parser = argparse.ArgumentParser(description='parse DRM mutations')
    parser.add_argument('-a', '--haplotypes', dest='haps',
                        help='file with amino acid haplotypes')

    args = parser.parse_args()
    return args


def aa_unpack(mut_string):
    '''Helper function used to extract a set of amminoacids from the
    merged csv files.'''
    if not mut_string.startswith('NOT'):
        return set(mut_string)
    return aa_set - set(mut_string.split()[1])


def parse_merged(mer_file):
    '''This is done by hand because it was too complicated to achieve
    this functionality with panda alone'''

    with open(mer_file) as csvfile:
        reader = csv.DictReader(csvfile)
        mdf = pd.DataFrame(columns=reader.fieldnames)
        for row in reader:
            row['freq'] = float(row['freq'])
            row['pos'] = int(row['pos'])
            if row['mut_x'] in aa_unpack(row['mut_y']):
                mdf = mdf.append(row, ignore_index=True)
            elif row['mut_x'] == '-' and 'd' in aa_unpack(row['mut_y']):
                mdf = mdf.append(row, ignore_index=True)
            elif row['mut_x'] and row['mut_y'] == '':
                row['category'] = 'unannotated'
                mdf = mdf.append(row, ignore_index=True)
    return mdf


def main(org=None, mut_file='annotated_mutations.csv',
         subtypes_file='subtype_evidence.csv'):
    '''What does the main do?'''

    import subprocess

    rh = open('report.md', 'w')
    write_header(rh, subtypes_file, org)

    print('Reading mutations from %s' % mut_file, file=sys.stderr)
    mutation_detected = pd.read_csv(mut_file)
    print('Shape is: ', mutation_detected.shape)

    if org == 'HIV':
        print('Parsing DRM from database', file=sys.stderr)
        resistance_mutations = parse_drm()
        print('Shape is: ', resistance_mutations.shape)

        mpd = pd.merge(mutation_detected, resistance_mutations, how='left',
                       on=['gene', 'pos'])
        mpd.to_csv(path_or_buf='merged_muts_drm_annotated.csv')
        print('Shape of raw merged is: ', mpd.shape)

        # too complicated with panda, do it by hand
        drms = parse_merged('merged_muts_drm_annotated.csv')
        print('Shape of merged is: ', drms.shape)
        #os.remove('merged_muts.csv')

        drms.drop(['', 'commented', 'mut_y'], axis=1, inplace=True)
        drms.rename(columns={'mut_x': 'mut'}, inplace=True)

        cols = ['gene', 'pos', 'mut', 'freq', 'category']
        drms = drms[cols]
        drms.sort_values(by=['gene', 'pos', 'freq'], inplace=True,
                         ascending=[True, True, False])
        drms.to_csv('annotated_DRM.csv', index=False)

        for gene in ['protease', 'RT', 'integrase']:
            gene_muts = drms[drms.gene == gene]
            if gene_muts.shape[0] == 0:
                print('No mutations on ', gene, file=sys.stderr)
                print('No mutations on %s' % gene, file=rh)
                print('-' * (len(gene) + 16), file=rh)
                print(file=rh)
                continue
            # sort was lost because
            gene_muts = gene_muts.sort_values(
                by=['gene', 'pos', 'freq'], ascending=[True, True, False])
            print('%s' % gene, file=rh)
            print('-'*len(gene), file=rh)
            h1 = '| position | mutation | frequency [%] |      category      |'
            print(h1, file=rh)
            h2 = '|:{:-^8}:|:{:-^8}:|:{:-^13}:|:{:-^18}:|'
            print(h2.format('', '', '', ''), file=rh)
            for index, row in gene_muts.iterrows():
                # # same pos mut tuple must give same annota, probably redundant
                # assert group['category'].nunique() == 1, \
                #     group['category'].describe()
                mut_cat = row['category']
                int_freq = int(round(100 * row['freq'], 0))
                print(
                    '|{: ^10}|{: ^10}|{: ^15}|{: ^20}|'.format(int(row['pos']),
                                                               row['mut'],
                                                               int_freq,
                                                               mut_cat),
                    file=rh)
            print('\n', file=rh)
    elif org == 'HCV':
        drms = pd.read_csv('annotated_mutations.csv')
        print('Shape of annotated_mutations is: ', mutation_detected.shape)

        if drms.shape[0] == 0:
            print('No mutations found', file=rh)
            print('-' * 26, file=rh)
            print('\n', file=rh)

        for gene in ['unknown']:
            gene_muts = drms[drms.gene == gene]
            if gene_muts.shape[0] == 0:
                print('No mutations on ', gene, file=sys.stderr)
                print('No mutations on %s' % gene, file=rh)
                print('-' * (len(gene) + 16), file=rh)
                print(file=rh)
                continue

            print('%s' % gene, file=rh)
            print('-'*len(gene), file=rh)
            h1 = '| position | mutation | frequency [%] |'
            print(h1, file=rh)
            h2 = '|:{:-^8}:|:{:-^8}:|:{:-^13}:|'
            print(h2.format('', '', ''), file=rh)
            for index, row in gene_muts.iterrows():
                int_freq = int(round(100 * row['freq'], 0))
                print(
                    '|{: ^10}|{: ^10}|{: ^15}|'.format(int(row['pos']),
                                                       row['mut'],
                                                       int_freq),
                    file=rh)
                del index
            print('\n', file=rh)

    rh.close()
    # convert to PDF with pandoc
    tmpl_file = os.path.abspath(os.path.join(db_dir, 'template.tex'))
    pand_cml = 'pandoc --template={} report.md -o report.pdf'.format(tmpl_file)
    print(pand_cml, file=sys.stderr)
    subprocess.call(pand_cml, shell=True)


if __name__ == '__main__':
    #args = parse_com_line()
    main(org=sys.argv[1], mut_file='annotated_mutations.csv',
         subtypes_file='subtype_evidence.csv')
